core/commonMain/src/kotlinx/serialization/SerializationExceptions.kt

```diff
@@ -57,15 +59,73 @@ public open class SerializationException : IllegalArgumentException {
 }
 
 /**
- * Thrown when [KSerializer] did not receive property from [Decoder], and this property was not optional.
+ * Thrown when [KSerializer] did not receive a non-optonal property from [CompositeDecoder] and [CompositeDecoder.decodeElementIndex]
+ * had already returned [CompositeDecoder.DECODE_DONE].
+ *
+ * [MissingFieldException] is thrown on missing field from all [auto-generated][Serializable] serializers and it
+ * is recommended to throw this exception from user-defined serializers.
+ *
+ * @see SerializationException
+ * @see KSerializer
  */
-@PublishedApi
-internal class MissingFieldException
-// This constructor is used by coroutines exception recovery
-internal constructor(message: String?, cause: Throwable?) : SerializationException(message, cause) {
-    // This constructor is used by the generated serializers
-    constructor(fieldName: String) : this("Field '$fieldName' is required, but it was missing", null)
-    internal constructor(fieldNames: List<String>, serialName: String) : this(if (fieldNames.size == 1) "Field '${fieldNames[0]}' is required for type with serial name '$serialName', but it was missing" else "Fields $fieldNames are required for type with serial name '$serialName', but they were missing", null)
+@ExperimentalSerializationApi
+public class MissingFieldException
```

Comment:
If we made the `JvmMissingFieldsExceptionTest` public, then we can refine some tests (for example) so that they use `assertFailsWith(JvmMissingFieldsExceptionTest::class, null, block)` and check the missing fields explicitly with `missingFields` property.

Reply:
The golden mask check should work for all IR backends. Probably, this test was written when JS/IR was still unstable.